src/dyad/core/dyad_core.c

```diff
@@ -934,58 +931,71 @@ dyad_rc_t dyad_consume (dyad_ctx_t* ctx, const char* fname)
         dyad_release_flock (ctx, fd, &exclusive_lock);
         goto consume_close;
     }
-    if ((file_size = get_file_size (fd)) <= 0) {
-        DYAD_LOG_INFO (ctx, "[node %u rank %u pid %d] File (%s with fd %d) is not fetched yet", \
-                       ctx->node_idx, ctx->rank, ctx->pid, fname, fd);
-        // Call dyad_fetch to get (and possibly wait on)
-        // data from the Flux KVS
-        rc = dyad_fetch (ctx, fname, &mdata);
-        // If an error occured in dyad_fetch, log an error
-        // and return the corresponding DYAD return code
-        if (DYAD_IS_ERROR (rc)) {
-            DYAD_LOG_ERROR (ctx, "dyad_fetch failed!\n");
-            dyad_release_flock (ctx, fd, &exclusive_lock);
-            goto consume_close;
-        }
-        // If dyad_fetch was successful, but resp is still NULL,
-        // then we need to skip data transfer.
-        // This will most likely happend because shared_storage
-        // is enabled
-        if (mdata == NULL) {
-            DYAD_LOG_INFO (ctx, "File '%s' is local!\n", fname);
-            rc = DYAD_RC_OK;
-            dyad_release_flock (ctx, fd, &exclusive_lock);
-            goto consume_close;
+    file_size = get_file_size (fd);
+    if (ctx->shared_storage) {
+        dyad_release_flock (ctx, fd, &exclusive_lock);
+        if (!ctx->use_fs_locks || file_size <= 0) {
+            // as file size was zero that means consumer won the lock first so has to wait for kvs.
+            // or we are using c++ streams.
```

Comment:
Clarify the comment by changing the following:
"or we are using c++ streams." => "or we cannot use file lock based synchronization as it does not work with the files managed by c++ fstream."

Reply:
FIXED